File: javascript/queen-attack/queen-attack.js
String.prototype.replace = function(index, new_chr) {
    return this.substr(0, index) + new_chr + this.substr(index + new_chr.length);
};

function Queens(posObj) {
	if ('undefined' === typeof posObj) {
		this.white = [0, 3];
		this.black = [7, 3];
	} else {
		if (posObj.white === posObj.black) {
			throw 'Queens cannot share the same space';
		}
		
		this.white = posObj.white;
		this.black = posObj.black;
	}
}

Queens.prototype.toString = function() {
	var stringified_board = "";
    for (var i = 0; i < 8; ++i) {
        stringified_board += "_ _ _ _ _ _ _ _\n";
    }
    
    var white_position = 2 * ((this.white[0] * 8) + this.white[1]);
    var black_position = 2 * ((this.black[0] * 8) + this.black[1]);

    return stringified_board.replace(white_position, 'W').replace(black_position, 'B');
};

Queens.prototype.canAttack = function() {
	// Check if they are in the same line
	if (this.white[0] === this.black[0]) {
		return true;
	}
	
	// Check if they are in the same column
	if (this.white[1] === this.black[1]) {
		return true;
	}
	
	// Check if they are in the same diagonal
	var deltax = this.white[0] - this.black[0];
	var deltay = this.white[1] - this.black[1];

	if (deltay === -deltax) {
		return true;
	}
	
	return deltay === deltax;
};

module.exports = Queens;

File: python/queen-attack/queen_attack.py
def init_matrix():
    return [['_' for j in range(8)] for i in range(8)]

def place_queens(white_queen, black_queen, board):
    board[white_queen[0]][white_queen[1]] = "W"
    board[black_queen[0]][black_queen[1]] = "B"

    return board

def valid_queen_position(queen):
    for coordinate in queen:
        if coordinate < 0 or coordinate > 7:
            return False

    return True

def queen_overlap(white, black):
    return white[0] == black[0] and white[1] == black[1]

def valid_queens(white, black):
    if queen_overlap(white, black):
        return False

    return valid_queen_position(white) and valid_queen_position(black)

def board(white_queen, black_queen):
    if not valid_queens(white_queen, black_queen):
        raise ValueError("Your queens got problems.")

    return [''.join(row) for row in place_queens(white_queen, black_queen, init_matrix())]

def can_attack(white, black):
    if not valid_queens(white, black):
        raise ValueError("Your queens got problems.")

    # Check if they are in the same line
    if white[0] == black[0]:
        return True

    # Check if they are in the same column
    if white[1] == black[1]:
        return True

    # Check if they are in the same diagonal
    deltax = white[0] - black[0]
    deltay = white[1] - black[1]

    if deltay == -deltax:
        return True

    if deltay == deltax:
        return True

    return False